### PR TITLE
Change "augmentation library" wording to just "augmentation", other updates

### PR DESCRIPTION
**Pull request by jakemac53**
A variety of proposed changes related to https://github.com/dart-lang/language/issues/3577.

- Changes references to 'augmentation library' to 'library augmentation', it isn't really a library on its own
- Changes 'library augment' => 'augment library' (reads better)

### CODE REVIEW

File: working/augmentation-libraries/feature-specification.md [hunk 1]
```diff
@@ -142,59 +141,58 @@ It is a compile-time error if:
...
 library.
 
 ### Merge order
 
 A library may apply multiple augmentations to itself. Also, augmentation files
-may themselves contain `import augment` directives. The entire tree of
-augmentations is recursively applied to the main library. The merge order is
-defined as a depth-first pre-order traversal of the `import augment` directives
-starting at the main library.
+may themselves contain `augment` directives. The entire tree of augmentations is
```
Comment (lrhn) [inline working/augmentation-libraries/feature-specification.md:181]: "`augment` directives" have something of the same issue as plain "augmentation", because I'm pretty certainly going to use *something like that* that to cover `augment class` declarations too, which an augmentation file can also contain.
If the distinction is that the "augment import" is a *directive* and the other ones are *declarations*, I don't think that's going to be distinct enough. 
At least not unless we make that distinction *very explicit* where we introduce the directive above.

Say, something like:
> A directive of the form `augment "file.dart";` which applies a library augmentation to the library of the directive, is called an _`augment` directive_, which distinguishes it from _`augment` declarations_, which are declarations that an augmentation file can also contain, and whose syntax also starts with `augment`. These `augment` declarations will be introduced later.

If we do that, very up front, I think we *can* use "`augment` directive" unambiguously.
(It's really about defining the term. If we do nothing, "`augment` directive" is just a descriptive term, not a declared concept, and people can interpret it, and interpret it incorrectly. If we *define* the meaning, then the term has only that one meaning in this document. I'm a big fan of giving things defined names.)

File: working/augmentation-libraries/feature-specification.md [hunk 2]
```diff
@@ -142,59 +141,58 @@ It is a compile-time error if:
...
+recursively applied to the main library. The merge order is defined as a
+depth-first pre-order traversal of the `augment` directives starting at the main
```
Comment (lrhn) [inline working/augmentation-libraries/feature-specification.md:183]: of the augmentation files, in the source order of their `augment` directives,

File: working/augmentation-libraries/feature-specification.md [hunk 3]
```diff
@@ -142,59 +141,58 @@ It is a compile-time error if:
...
+library
 
-Within a single augmentation library, you may augment the same declaration
-multiple times, whether they are top level or nested declarations. The merge
-order of these is defined as the source order of the augmentations.
+Within a single augmentation file, you may augment the same declaration multiple
+times, whether they are top level or nested declarations. The merge order of
```
Comment (lrhn) [inline working/augmentation-libraries/feature-specification.md:187]: Nit: "they" should refer to a plural object, but I think it refers to "same declaration", so "it is" (or better "the declaration is", unqualified "it"s risk being ambiguous.)

File: working/augmentation-libraries/feature-specification.md [hunk 4]
```diff
@@ -142,59 +141,58 @@ It is a compile-time error if:
...
+these is defined as the source order of the augmentations.
```
Comment (lrhn) [inline working/augmentation-libraries/feature-specification.md:188]: of the agumentation declarations. 

(Source order imples syntax, declarations are syntax. And the unqualified "augmentations" here is an example of what would be ambiguous if "augmentation" was defined as a library augmentation.)

File: working/augmentation-libraries/feature-specification.md [hunk 5]
```diff
@@ -301,7 +299,7 @@ It is a compile-time error if:
 
 *   An augmenting declaration appears in a library before the library where the
     original declaration occurs, according to merge order. *An augmentation
-    library can both declare a new declaration and augment it in the same file.*
+    can both declare a new declaration and augment it in the same file.*
```
Comment (lrhn) [inline working/augmentation-libraries/feature-specification.md:302]: (Can it augment the same member twice in the same `augment class`? ... I know, I should just check. Should it?)

File: working/augmentation-libraries/feature-specification.md [hunk 6]
```diff
@@ -406,7 +404,7 @@ It is a compile-time error if:
     doesn't accomplish anything semantically. But it ensures that anyone reading
     the augmenting type can see the declarations of any type parameters that it
     uses in its body and avoids potential confusion with other top-level
-    variables that might be in scope in the augmentation library.*
+    variables that might be in scope in the augmentation.*
```
Comment (lrhn) [inline working/augmentation-libraries/feature-specification.md:407]: augmentation file.

File: working/augmentation-libraries/feature-specification.md [hunk 7]
```diff
@@ -494,14 +492,14 @@ replaces the variable's implicit getter body with the augmenting getter's.
...
+    in the main library. Inside the augmenting body, an `augmented` expression
+    invokes the original getter.
 
-*   **Augmenting with a setter:** A setter in an augmentation library can
-    augment a setter in the main library or the implicit setter defined by a
-    non-final variable in the main library. Inside the augmenting setter, an
+*   **Augmenting with a setter:** A setter in an augmentation can augment a
+    setter in the main library or the implicit setter defined by a non-final
+    variable in the main library. Inside the augmenting setter, an
     `augmented =` expression invokes the original setter.
```
Comment (lrhn) [inline working/augmentation-libraries/feature-specification.md:503]: Drive-by: Consider showing the operand: 
> <code>augmented = _expr_</code> expression ...

File: working/augmentation-libraries/feature-specification.md [hunk 8]
```diff
@@ -573,8 +571,8 @@ It is a compile-time error if:
     a variable with an initializer.
 
 *   A final variable is augmented with a setter. (Instead, the augmentation
-    library can declare a *non-augmenting* setter that goes alongside the
-    implicit getter defined by the final variable.)
+    can declare a *non-augmenting* setter that goes alongside the implicit
+    getter defined by the final variable.)
```
Comment (lrhn) [inline working/augmentation-libraries/feature-specification.md:575]: (This is the general rule of "you can't augment what doesn't exist", and a final variable doesn't introduce a setter, so an `augment` setter declaration cannot augment what isn't there. Right?)

File: working/augmentation-libraries/feature-specification.md [hunk 9]
```diff
@@ -599,40 +597,41 @@ An augmenting enum value is allowed to invoke a different constructor than
 the original enum value, or provide an argument list where none was present
 before.
 
-New enum values may also be defined in the augmenting library, and they will
-be appended to the original values in source and augmentation traversal order.
+New enum values may also be defined in the augmentation, and they will be
+appended to the original values in source and augmentation traversal order.
```
Comment (lrhn) [inline working/augmentation-libraries/feature-specification.md:601]: (I'd try to define it so that there is only one order, which is source order within a file, then recursive traverseal of the augmentation files of `augment` directives in the current file, in their source order. Then this is just "in augmentation traversal order", the source order within a file is included in that. Otherwise pedants might start asking what the priority is between "source" and "augmentation traversal" order. They do that.)

File: working/augmentation-libraries/feature-specification.md [hunk 10]
```diff
@@ -710,13 +709,13 @@ top-level scope where declarations are defined. They also share a single private
 namespace. This means that private declarations in the main library or an
 augmentation are visible to all augmentations.
 
-Unlike part files, an augmentation library has its own import scope surrounding
+Unlike part files, an augmentation has its own import scope surrounding that
```
Comment (lrhn) [inline working/augmentation-libraries/feature-specification.md:712]: augmentation file

also: "that \\n that" -> "that"

File: working/augmentation-libraries/feature-specification.md [hunk 11]
```diff
@@ -710,13 +709,13 @@ top-level scope where declarations are defined. They also share a single private
...
 that shared top-level scope. Any libraries the augmentation imports are visible
```
Comment (lrhn) [inline working/augmentation-libraries/feature-specification.md:713]: the augmentation file imports
or
the library augmentation imports

(imports can be treated both syntactically or semantically, either works)

File: working/augmentation-libraries/feature-specification.md [hunk 12]
```diff
@@ -752,8 +751,8 @@ augment class C {
 ```
 
 This code is fine. Code in C in the main library can refer to members added in
```
Comment (lrhn) [inline working/augmentation-libraries/feature-specification.md:753]: "in C" -> "in the class `C`"

(Code quote source names, and make it clear that there is no C (the language) code here.)

File: working/augmentation-libraries/feature-specification.md [hunk 13]
```diff
@@ -1,9 +1,9 @@
-# Augmentation Libraries
+# Augmentations
 
 Author: rnystrom@google.com, jakemac@google.com
 Version: 1.13 (see [Changelog](#Changelog) at end)
 
-Augmentation libraries allow splitting a Dart library into files. Unlike part
+Augmentations allow splitting a Dart library into files. Unlike part
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/augmentation-libraries/feature-specification.md:6]: Yeah, I think I'm with Lasse on this. Part of why I called these "augmentation libraries" is to avoid the ambiguity of "augmentation" on its own.

Initially, I called them "augmentation libraries" to emphasize that (unlike part files), they are library-like in that they can have their own imports. This is less true now that they share a private scope with the main library, so removing "library" from the name probably makes sense.

Maybe "library augmentation"? So the file itself isn't a library, but it is a thing that augments a library.

File: working/augmentation-libraries/feature-specification.md [hunk 14]
```diff
@@ -115,7 +114,7 @@ library. Augmentations are distinguished by a special `library` directive with
 an `augment` modifier, like so:
 
 ```dart
-library augment 'main_library.dart';
+augment library 'main_library.dart';
```
Comment (munificent) [inline working/augmentation-libraries/feature-specification.md:117]: For what it's worth, my thinking with the original syntax was that you would read `library augment 'foo.dart'` as "a `library` that `augment`s `'foo.dart'`".

But I like the syntax you have here much better. I like that it parallels `augment` as a modifier on other declarations where `augment` precedes the thing it augments.

File: working/augmentation-libraries/feature-specification.md [hunk 15]
```diff
@@ -142,59 +141,58 @@ It is a compile-time error if:
...
+### Applying an augmentation
 
-A main library applies an augmentation to itself using a new import directive
-that looks like this:
+A main library applies an augmentation to itself using a new `augment` directive
+which looks like this:
 
 ```dart
-import augment 'some_augmentation.dart';
+augment 'some_augmentation.dart';
```
Comment (munificent) [inline working/augmentation-libraries/feature-specification.md:152]: This reads strangely to me now because in all the other places where we use `augment` it means that the subsequent thing is the thing *being augmented* but here it means *thing doing the augmenting*.

I don't have any great suggestions, but maybe keep `import augment` because that might be a little clearer? Perhaps `with augment`?

**Review by munificent — CHANGES_REQUESTED** [review]
I like changing `library augment` to `augment library`. I'm not as keen on changing `import augment` to just `augment` and I think using "augmentation" as a standalone noun will end up confusing.

But, overall, I think this is a step in the right direction. Thanks for taking my strawman syntax and beating it into shape. :)

File: working/augmentation-libraries/feature-specification.md (hunk 15 above)
Comment (jakemac53) [reply working/augmentation-libraries/feature-specification.md:152]: Yes I agree, this is the one part I don't like now, for the same reasons. I suggested "augmentation" above for this but then its yet another special word to introduce.

Comment (munificent) [reply working/augmentation-libraries/feature-specification.md:152]: It would be nice to not have to introduce every conjugation of "augment" into the language: `augment`, `augments`, `augmentation`, `augmenting`, `augmentable`, `augmenter`, etc.. :)

I think `import augment` works OK.

Comment (jakemac53) [reply working/augmentation-libraries/feature-specification.md:152]: Fwiw, to me this is the strongest argument for just re-using the "part" directive 🤣. But, we would be breaking the semantics of it, so I think that is a no-go. But, we could just be bold...

Comment (munificent) [reply working/augmentation-libraries/feature-specification.md:152]: > Fwiw, to me this is the strongest argument for just re-using the "part" directive 🤣.

I would *love* to repurpose part files for this feature. "Augment" works OK as a word, but it's not exactly common and is pretty hard for ESL speakers from what I've gathered.

> But, we would be breaking the semantics of it, so I think that is a no-go. But, we could just be bold...

Yeah, it's probably a bridge too far. :( From a quick scrape of a pub package corpus:

```
-- File (56652 total) --
  45719 ( 80.701%): no part files    ===================================
   7759 ( 13.696%): is part file     ======
   3174 (  5.603%): uses part files  ===
```

So part files are actually pretty common. I suspect almost all of those are generated code?

Comment (jakemac53) [reply working/augmentation-libraries/feature-specification.md:152]: > So part files are actually pretty common. I suspect almost all of those are generated code?

Could you check how many of these files have multiple extensions, as a proxy for generated files (*.*.dart)?

Comment (munificent) [reply working/augmentation-libraries/feature-specification.md:152]: From skimming the file paths as they blew by on my terminal, I didn't notice a lot of extensions that suggested they were generated.

Comment (lrhn) [reply working/augmentation-libraries/feature-specification.md:152]: We could use `part` here, and then whether the part file starts with `part of ....` or `augment library ...` decides whether it can have imports or not. Whether it's a simple part file or a library augmentation part file. (But if we allow `augment` in every file, #3576, we no longer have library augmentations, we just have files which may or may not include member augmentation. Then it's just about whether that file allows imports or it inherits the parent file's imports. At which point the inclusion syntax shouldn't mention `augment`, and I don't know which name would fit so it could be a `*word* part of ...` vs `part of ....`. )

**Comment by jakemac53** [comment]
> @jakemac53 what's the status of these changes? I want to update the textmate grammar (and LSP semantic tokens) but I was waiting for this to ensure any tests I add are using the correct code (definitely not urgent, just trying to clear some things off my list :-))

The status is I am overwhelmed 🤣 . I will try to circle back here soon though, thanks for the ping.

File: working/augmentation-libraries/feature-specification.md [hunk 16]
```diff
@@ -820,15 +819,15 @@ libraryAugmentImport ::= metadata 'import' 'augment' uri ';'
 ```
 
 A library directive may contain `augment` followed by a URI to denote the file
-as an augmentation library:
+as an augmentation:
```
[the remark this answers is not in the document]
Comment (jakemac53) [reply working/augmentation-libraries/feature-specification.md:826]: Added a TODO, I want to tackle this separately.

File: working/augmentation-libraries/feature-specification.md [hunk 17]
```diff
@@ -903,11 +902,10 @@ To apply an augmentation to the main library:
 1.  Merge the augmentation's declarations into the main library's top-level
     namespace using the following procedure.
 
-1.  For each `import augment` directive in the augmentation library, in
-    syntactic order:
+1.  For each `augment` directive in the augmentation, in syntactic order:
```
[the remark this answers is not in the document]
Comment (jakemac53) [reply working/augmentation-libraries/feature-specification.md:911]: That reads weird I think, I made the change above though, which helps.

File: working/augmentation-libraries/feature-specification.md [hunk 18]
```diff
@@ -66,20 +66,19 @@ part of macro expansion, it's still useful to have it be in a canonical
...
 into smaller files.
 
-## Augmentation libraries
+## Augmentations
 
-An augmentation library is a separate file that *augments* another *main*
-library. An augmentation is somewhere between a separate library and a part
-file.
+An augmentation is a separate file that *augments* another *main* library. An
+augmentation is somewhere between a separate library and a part file.
```
[the remark this answers is not in the document]
Comment (jakemac53) [reply working/augmentation-libraries/feature-specification.md]: I reworked this entirely

**Comment by jakemac53** [comment]
I pushed some changes based on the feedback here but didn't get to it all yet, will try to finish up tomorrow.

File: working/augmentation-libraries/feature-specification.md [hunk 19]
```diff
@@ -1,12 +1,12 @@
...
-Augmentations allow splitting a Dart library into files. Unlike part
-files, each augmentation has its [own imports][part imports] and top-level
-scope. Augmentations can add new top-level declarations, inject new members into
-classes, and wrap functions and variables in additional code.
+Augmentations allow spreading your implementation across multiple locations,
+both within a single file and across multiple files. They can add new top-level
+declarations, inject new members into classes, and wrap functions and variables
+in additional code.
 
 [part imports]: https://github.com/dart-lang/language/issues/519
```
Comment (munificent) [inline working/augmentation-libraries/feature-specification.md:11]: You can remove this link reference if you're removing the bit about part files from the intro.

File: working/augmentation-libraries/feature-specification.md [hunk 20]
```diff
@@ -66,51 +66,51 @@ part of macro expansion, it's still useful to have it be in a canonical
...
-    library, or provide a body if none was present.
+*   Function augmentations, which can replace the body of a function, or provide
+    a body if none was present.
 
-*   A variable in the augmentation may wrap the initializer of a variable in the
-    main library, or provide an initializer if none was present.
-
-*   An enum value in the augmentation may replace the argument list of an enum
-    value in the main library, or provide an argument list if none was present.
+*   Variable augmentations may wrap the initializer of a variable in the
```
Comment (munificent) [inline working/augmentation-libraries/feature-specification.md:105]: Because I'm a tech writing nerd now and list bullets lists to have parallel grammar structure:

> Variable augmentations, which can wrap...

File: working/augmentation-libraries/feature-specification.md [hunk 21]
```diff
@@ -66,51 +66,51 @@ part of macro expansion, it's still useful to have it be in a canonical
...
+    augmented library, or provide an initializer if none was present.
 
 These can't be expressed today using only imports, exports, and part files.
 
-### Defining an augmentation
+### Defining an library augmentation
```
Comment (munificent) [inline working/augmentation-libraries/feature-specification.md:110]: "an" -> "a".

File: working/augmentation-libraries/feature-specification.md [hunk 22]
```diff
@@ -66,51 +66,51 @@ part of macro expansion, it's still useful to have it be in a canonical
...
 
-An augmentation has almost the same syntax and semantics as a normal Dart
-library. Augmentations are distinguished by a special `library` directive with
+An library augmentation has almost the same syntax and semantics as a normal
```
Comment (munificent) [inline working/augmentation-libraries/feature-specification.md:112]: Here too.

File: working/augmentation-libraries/feature-specification.md [hunk 23]
```diff
@@ -122,30 +122,34 @@ syntax][import]?**
 
 [import]: https://github.com/dart-lang/language/blob/master/working/0649%20-%20Import%20shorthand/proposal.md
 
-The URL points to the main library that this augmentation is applied to.
+The URL points to the augmented library that this augmentation is applied to,
+which must be a regular library (and not a library augmentation).
```
Comment (munificent) [inline working/augmentation-libraries/feature-specification.md:126]: The parenthetical implies that a hand-authored library augmentation can't contain macro applications. Is that deliberate?

The "Merge Order" section still says that an entire tree of library augmentations is possible. Or is it that every library augmentation must point all the way back to the augmented library and not any of the library augmentations?

File: working/augmentation-libraries/feature-specification.md [hunk 24]
```diff
@@ -122,30 +122,34 @@ syntax][import]?**
...
+*   A library augmentation contains a normal `library` directive. They are not
+    self-contained libraries, only pieces of the augmented library.
+
 ### Applying an augmentation
 
-A main library applies an augmentation to itself using a new `augment` directive
+A library applies an augmentation to itself using a new `augment` directive
 which looks like this:
 
 ```dart
```
Comment (munificent) [inline working/augmentation-libraries/feature-specification.md:155]: I'm iffy on this syntax. It reads to me like it's saying that *this* file is augmenting `'some_augmentation.dart'` when it's the exact opposite.

This used to be `import augment`, right? I think I still lean towards that syntax.

File: working/augmentation-libraries/feature-specification.md [hunk 25]
```diff
@@ -239,12 +242,13 @@ augment void trace() {
...
-declarations in those libraries&mdash;new declarations or augmentations&mdash;
-are processed in that order, and source order within that.
+declarations in those library augmentations
+&mdash;new declarations or augmentations&mdash; are processed in that order,
+and source order within that.
 
 This order is user-visible in two ways:
 
-*   A non-augmenting declaration must appear first before it can be augmented.
+*   A regular declaration must appear first before it can be augmented.
```
Comment (munificent) [inline working/augmentation-libraries/feature-specification.md:251]: Nit: I do like "non-augmenting" here because it's a little less vague than "regular", which isn't explicitly defined.

Or maybe do:

> A regular (i.e. not augmenting)...

File: working/augmentation-libraries/feature-specification.md [hunk 26]
```diff
@@ -785,45 +789,47 @@ some_lib.dart       | some_augment.dart
...
-3.  Resolve and type-check the main library and all of its augmentations now
-    that all type namespaces are complete.
+3.  Resolve and type-check the augmented library and all of its augmentations
+    now that all type namespaces are complete.
 
 ## Syntax
 
-The grammar changes are fairly simple. A main library can apply an augmentation
-using a new directive:
+The grammar changes are fairly simple. A library can apply an augmentation using
```
Comment (munificent) [inline working/augmentation-libraries/feature-specification.md:814]: Heh, I suppose this first sentence as a little too optimistic about simplicity. :)

File: working/augmentation-libraries/feature-specification.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (munificent) [reply working/augmentation-libraries/feature-specification.md:826]: I'll help out with the grammar changes.

File: working/augmentation-libraries/feature-specification.md [hunk 27]
```diff
@@ -930,26 +936,26 @@ To merge a set of declarations `D` into a namespace:
...
+            namespace of the augmented type.
 
     1.  Else, if the declaration is a function, getter, setter, or operator:
 
-        1.  Replace the body of the main function with the augmenting function's
-            body. Inside the augmenting body, a `super augment()`, `super
-            augment`, `super augment =`, or `super augment <op>` expression as
-            appropriate calls the original function body.
+        1.  Replace the body of the augmented function with the augmenting
+            function's body. Inside the augmenting body, a `augmented()`,
```
Comment (munificent) [inline working/augmentation-libraries/feature-specification.md:956]: Nit: "a `augmented()`" -> "an `augmented()`"

**Review by munificent — CHANGES_REQUESTED** [review]
I'd like to talk more about whether we want `augment` or `import augment` (and I'm sorry if we discussed that before I don't remember all the details now). Aside from that, this LGTM.

File: working/augmentation-libraries/feature-specification.md (hunk 24 above)
Comment (jakemac53) [reply working/augmentation-libraries/feature-specification.md:155]: Yeah, I will probably revert this part, as I agree it is a bit weird. I don't like either of them really, but I don't have an alternative proposal either 🤷‍♂️ .

File: working/augmentation-libraries/feature-specification.md (hunk 5 above)
Comment (jakemac53) [reply working/augmentation-libraries/feature-specification.md]: Yes it can, and we do this in the macro generated augmentations to make them more concise, if multiple macros augment the same member

File: working/augmentation-libraries/feature-specification.md (hunk 8 above)
Comment (jakemac53) [reply working/augmentation-libraries/feature-specification.md:581]: Yes, I don't think it hurts to explicitly call it out though? But that is the reasoning for sure.

File: working/augmentation-libraries/feature-specification.md (hunk 9 above)
Comment (jakemac53) [reply working/augmentation-libraries/feature-specification.md]: I changed to just "in augmentation traversal order", that is what you meant right? :)

**Comment by jakemac53** [comment]
Updated to change `augment <uri>;` back to `import augment <uri>;`, addressed the rest of the feedback, PTAL

**Review by munificent — APPROVED** [review]
This looks great!

**Comment by lrhn** [comment]
My only worry with `augment import` is that it's long.
There have been suggestions that it should replace `part` as the way to modularize a library. Then it might count against it that the syntax to include focuses on the augmentation, even if the file it referred to doesn't contain any `augment` declarations at all.

(But I guess this is me coming back to suggesting that we just use `part`, allow augmentation declarations in all files, allow import, export and part declarations in part files, and if a part file declares any import, then it doesn't get the parent file's imports. We don't need a new kind of file, another way to include declarations, we already have `part`.)

**Comment by jakemac53** [comment]
> My only worry with `augment import` is that it's long.

I agree, I do think we might want to come back to this, with some new keyword or by re-using `part`. So this isn't the end of that discussion, I just wasn't happy with a bare `augment` as the thing.

**Comment by jakemac53** [comment]
cc @johnniwinther @scheglov note that this change does presumably require some worth for the CFE/Analyzer and the code that generates augmentation libraries. I am not sure how we want to handle those changes in the best way.

**Comment by scheglov** [comment]
I like this change, it reads consistently with other augmentations.

https://dart-review.googlesource.com/c/sdk/+/358060 probably prepares `analyzer/` and `analysis_server/` for the parser change. I cannot land it, of course. So, it is to be merged with the parse changes.
